### PR TITLE
cpdp-838: implemented enter trn page

### PR DESCRIPTION
**Pull request by AndyHWChung**

### CODE REVIEW

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/models/bpr/UnsuccessfulNameMatchAttempt.scala [hunk 1]
```diff
@@ -16,19 +16,47 @@
...
+    implicit val trustNameMatchDetailsFormat: OFormat[TrustNameMatchDetails] = Json.format[TrustNameMatchDetails]
+
+  }
+
+
+
+//  @SuppressWarnings(Array("org.wartremover.warts.PublicInference"))
+// implicit def unsuccesfulNameMatchAttemptsFormat[A <: NameMatchDetails : OFormat]: OFormat[UnsuccessfulNameMatchAttempts[A]] =
+//    Json.format[UnsuccessfulNameMatchAttempts[A]]
+
```
Comment (co-category) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/models/bpr/UnsuccessfulNameMatchAttempt.scala]: Is this not required anymore? If not please remove

Comment (AndyHWChung) [reply app/uk/gov/hmrc/cgtpropertydisposalsfrontend/models/bpr/UnsuccessfulNameMatchAttempt.scala]: oh yeah, whoops - I'll get rid of that

Comment (AndyHWChung) [reply app/uk/gov/hmrc/cgtpropertydisposalsfrontend/models/bpr/UnsuccessfulNameMatchAttempt.scala]: it's gone now

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/models/bpr/UnsuccessfulNameMatchAttempt.scala [hunk 2]
```diff
@@ -16,19 +16,41 @@
...
+    implicit val eq: Eq[NameMatchDetails] = Eq.fromUniversalEquals
+
+    implicit val individualNameMatchDetailsFormat: OFormat[IndividualNameMatchDetails] = Json.format[IndividualNameMatchDetails]
+
+    implicit val trustNameMatchDetailsFormat: OFormat[TrustNameMatchDetails] = Json.format[TrustNameMatchDetails]
+
+  }
+
+  implicit def unsuccesfulNameMatchAttemptsReads[A <: NameMatchDetails : Reads]: Reads[UnsuccessfulNameMatchAttempts[A]] =
```
Comment (co-category) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/models/bpr/UnsuccessfulNameMatchAttempt.scala]: small typo on the implicits identifier: missing another `s` in identifier.

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/models/bpr/UnsuccessfulNameMatchAttempt.scala [hunk 3]
```diff
@@ -16,19 +16,41 @@
...
+    Json.reads[UnsuccessfulNameMatchAttempts[A]]
+
+  implicit def unsuccesfulNameMatchAttemptsWrites[A <: NameMatchDetails : Writes]: Writes[UnsuccessfulNameMatchAttempts[A]] =
```
Comment (co-category) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/models/bpr/UnsuccessfulNameMatchAttempt.scala]: same here

Comment (AndyHWChung) [reply app/uk/gov/hmrc/cgtpropertydisposalsfrontend/models/bpr/UnsuccessfulNameMatchAttempt.scala]: I've corrected that typo now

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/models/bpr/UnsuccessfulNameMatchAttempt.scala [hunk 4]
```diff
@@ -16,19 +16,41 @@
...
-  lastNameTried: IndividualName,
-  lastSAUTRTried: SAUTR
+  lastDetailsTried: A
 )
 
 object UnsuccessfulNameMatchAttempts {
 
-  implicit val format: OFormat[UnsuccessfulNameMatchAttempts] = Json.format
+  sealed trait NameMatchDetails extends Product with Serializable
+
```
Comment (co-category) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/models/bpr/UnsuccessfulNameMatchAttempt.scala]: I can see why you have gone down this approach but I think it is more complex than it needs to be for what it does. Essentially we just want to read the try-count from the cache for an individual or a trust. It seems that this type-level machinery could have been replaced by a simple method taking an enum to represent the entity type. Am I missing something?

**Pull request merged**